gluoncv/auto/estimators/ssd/utils.py

```diff
@@ -0,0 +1,146 @@
+        # coco validation is slow, consider increase the validation interval
+        if args.validation.val_interval == 1:
+            args.validation.val_interval = 10
+    else:
+        raise NotImplementedError('Dataset: {} not implemented.'.format(dataset))
+    return train_dataset, val_dataset, val_metric
+
+def _get_dataloader(net, train_dataset, val_dataset, data_shape, batch_size, num_workers, ctx):
+    """Get dataloader."""
+    width, height = data_shape, data_shape
+    # use fake data to generate fixed anchors for target generation
+    with autograd.train_mode():
+        _, _, anchors = net(mx.nd.zeros((1, 3, height, width), ctx))
+    anchors = anchors.as_in_context(mx.cpu())
+    batchify_fn = Tuple(Stack(), Stack(), Stack())  # stack image, cls_targets, box_targets
+    train_loader = gluon.data.DataLoader(
+        train_dataset.transform(SSDDefaultTrainTransform(width, height, anchors)),
+        batch_size, True, batchify_fn=batchify_fn, last_batch='rollover', num_workers=num_workers)
+    val_batchify_fn = Tuple(Stack(), Pad(pad_val=-1))
+    val_loader = gluon.data.DataLoader(
+        val_dataset.transform(SSDDefaultValTransform(width, height)),
+        batch_size, False, batchify_fn=val_batchify_fn, last_batch='keep', num_workers=num_workers)
+    return train_loader, val_loader
+
+def _get_dali_dataset(dataset_name, devices, args):
+    if dataset_name.lower() == "coco":
+        # training
+        expanded_file_root = os.path.expanduser(args.dataset_root)
+        coco_root = os.path.join(expanded_file_root,
+                                 'coco',
+                                 'train2017')
+        coco_annotations = os.path.join(expanded_file_root,
+                                        'coco',
+                                        'annotations',
+                                        'instances_train2017.json')
+        if args.horovod:
+            train_dataset = [gdata.COCODetectionDALI(num_shards=hvd.size(), shard_id=hvd.rank(), file_root=coco_root,
+                                                     annotations_file=coco_annotations, device_id=hvd.local_rank())]
+        else:
+            train_dataset = [gdata.COCODetectionDALI(num_shards=len(devices), shard_id=i, file_root=coco_root,
+                                                     annotations_file=coco_annotations, device_id=i) for i, _ in enumerate(devices)]
+
+        # validation
+        if (not args.horovod or hvd.rank() == 0):
+            val_dataset = gdata.COCODetection(root=os.path.join(args.dataset_root + '/coco'),
+                                              splits='instances_val2017',
+                                              skip_empty=False)
+            val_metric = COCODetectionMetric(
+                val_dataset, os.path.join(args.logdir, args.save_prefix + '_eval'), cleanup=True,
+                data_shape=(args.data_shape, args.data_shape))
```

Comment:
args.ssd.data_shape

Reply:
Thx Chongruo, the problems have been corrected.